File: windows/src/TokenBar/Helpers/RateLimitReset.cs
using System;
using System.Globalization;

namespace TokenBar.Helpers
{
    /// <summary>
    /// 统一解析各厂商 x-ratelimit-reset-* / anthropic-ratelimit-*-reset 头里的「距重置时长」。
    /// 与 mac 端 RateLimitReset 语义一致：
    ///   - Go duration："20ms"、"1s"、"6m0s"、"1h2m"、"1.5s"
    ///   - 纯秒数："60"、"0.5"
    ///   - unix 秒时间戳（&gt; 1e9 视为绝对时间，换算为距今秒数）
    ///   - unix 毫秒时间戳（&gt; 1e12）
    /// 结果 clamp 到 [0, 86400]；解析失败返回 null，由调用方决定兜底值。
    /// </summary>
    public static class RateLimitReset
    {
        public const double MaxSeconds = 86400;

        public static double? Parse(string? raw)
        {
            if (string.IsNullOrWhiteSpace(raw)) return null;
            var text = raw.Trim();

            // 1. 纯数字：秒 / unix 秒 / unix 毫秒
            if (double.TryParse(text, NumberStyles.Float, CultureInfo.InvariantCulture, out var number))
            {
                if (double.IsNaN(number) || double.IsInfinity(number)) return null;
                var nowSec = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0;
                double seconds;
                if (number > 1e12) seconds = number / 1000.0 - nowSec;        // 毫秒时间戳
                else if (number > 1e9) seconds = number - nowSec;             // 秒时间戳
                else seconds = number;
                return Clamp(seconds);
            }

            // 2. Go duration：依次读取 <数字><单位>
            double total = 0;
            var i = 0;
            var matchedAny = false;
            while (i < text.Length)
            {
                var start = i;
                while (i < text.Length && (char.IsDigit(text[i]) || text[i] == '.')) i++;
                if (start == i) return null; // 单位前必须有数字
                if (!double.TryParse(text.AsSpan(start, i - start), NumberStyles.Float, CultureInfo.InvariantCulture, out var value))
                    return null;

                var unitStart = i;
                while (i < text.Length && char.IsLetter(text[i])) i++;
                var unit = text.Substring(unitStart, i - unitStart).ToLowerInvariant();

                double factor = unit switch
                {
                    "ms" => 0.001,
                    "s" => 1,
                    "m" => 60,
                    "h" => 3600,
                    "d" => 86400,
                    "us" or "µs" or "ns" => 0, // 过小的单位按 0 计，仍视为合法
                    _ => double.NaN
                };
                if (double.IsNaN(factor)) return null;
                total += value * factor;
                matchedAny = true;
            }

            return matchedAny ? Clamp(total) : null;
        }

        private static double Clamp(double seconds)
        {
            if (double.IsNaN(seconds)) return 0;
            return Math.Clamp(seconds, 0, MaxSeconds);
        }
    }
}

File: mac/Sources/TokenBar/Services/RateLimitReset.swift
import Foundation

/// 各厂商 `x-ratelimit-reset-*` 头的统一解析。
///
/// 网关返回的形状五花八门：OpenAI 是 Go duration（`20ms` / `1s` / `6m0s` / `1h2m`），
/// 部分中转站是纯秒数（`30`），也有直接给 unix 时间戳的（秒 `1710000000` 或毫秒 `1710000000000`）。
/// 以前各服务各写一份，且把时间戳当秒数会算出几十年后的重置时间。
/// 这里按量级区分：> 1e12 视为毫秒时间戳，> 1e9 视为秒时间戳，其余视为时长；结果统一 clamp 到 [0, 24h]。
public enum RateLimitReset {
    public static let maxSeconds: TimeInterval = 86400

    /// 返回距今的秒数；无法解析返回 nil（调用方自行决定兜底值，不要再假装是 1 秒）。
    public static func parse(_ raw: String?, now: Date = Date()) -> TimeInterval? {
        guard let raw else { return nil }
        let clean = raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        if clean.isEmpty { return nil }

        if let direct = Double(clean) {
            if direct > 1e12 { return clamp(Date(timeIntervalSince1970: direct / 1000).timeIntervalSince(now)) }
            if direct > 1e9 { return clamp(Date(timeIntervalSince1970: direct).timeIntervalSince(now)) }
            return clamp(direct)
        }

        // Go duration：数字后跟单位，可多段拼接
        var total: TimeInterval = 0
        var number = ""
        var unit = ""
        var sawUnit = false
        func flush() -> Bool {
            guard let value = Double(number) else { return false }
            switch unit {
            case "ms": total += value / 1000
            case "s": total += value
            case "m": total += value * 60
            case "h": total += value * 3600
            case "d": total += value * 86400
            default: return false
            }
            sawUnit = true
            number = ""; unit = ""
            return true
        }
        for ch in clean {
            if ch.isNumber || ch == "." {
                if !unit.isEmpty { guard flush() else { return nil } }
                number.append(ch)
            } else if ch.isLetter {
                unit.append(ch)
            } else if ch == " " {
                continue
            } else {
                return nil
            }
        }
        if !number.isEmpty || !unit.isEmpty { guard flush() else { return nil } }
        return sawUnit ? clamp(total) : nil
    }

    private static func clamp(_ v: TimeInterval) -> TimeInterval {
        min(max(v, 0), maxSeconds)
    }
}
